File: app/src/main/java/org/zeroxlab/momodict/archive/FileSet.java
package org.zeroxlab.momodict.archive;

import java.util.HashMap;
import java.util.Map;

import rx.functions.Action0;

import static org.zeroxlab.momodict.archive.FileSet.Type.DICT;
import static org.zeroxlab.momodict.archive.FileSet.Type.IDX;
import static org.zeroxlab.momodict.archive.FileSet.Type.IFO;

/**
 * A structure for extracted files path.
 */
public class FileSet {

    private Map<Type, String> mPaths = new HashMap<>();
    private Action0 mCleanCallback;

    public enum Type {
        IFO,
        IDX,
        DICT,
    }

    /**
     * For a dictionary, to check every necessary file exists.
     * @return true if necessary files exist.
     */
    public boolean isSane() {
        return has(IFO) && has(IDX) && has(DICT);
    }

    public boolean has(Type type) {
        return mPaths.containsKey(type);
    }

    public void set(Type type, String path) {
        mPaths.put(type, path);
    }

    public String get(Type type) {
        return mPaths.get(type);
    }

    public void setCleanCallback(Action0 cb) {
        mCleanCallback = cb;
    }

    /**
     * Call this to clean extracted files.
     */
    public void clean() {
        if (mCleanCallback != null) {
            mCleanCallback.call();
        }
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Type t : Type.values()) {
            sb.append(String.format("Type: %d, path: ", t, mPaths.get(t)));
        }
        return sb.toString();
    }
}

File: app/src/main/java/org/zeroxlab/momodict/archive/FileSet.kt
package org.zeroxlab.momodict.archive

import org.zeroxlab.momodict.archive.FileSet.Type.DICT
import org.zeroxlab.momodict.archive.FileSet.Type.IDX
import org.zeroxlab.momodict.archive.FileSet.Type.IFO
import java.util.HashMap

/**
 * A structure for extracted files path.
 */
class FileSet {

    private val mPaths = HashMap<Type, String>()
    private var mCleanCallback: (() -> Unit)? = null

    /**
     * For a dictionary, to check every necessary file exists.
     * @return true if necessary files exist.
     */
    val isSane: Boolean
        get() = has(IFO) && has(IDX) && has(DICT)

    enum class Type {
        IFO,
        IDX,
        DICT
    }

    fun has(type: Type): Boolean {
        return mPaths.containsKey(type)
    }

    operator fun set(type: Type, path: String) {
        mPaths[type] = path
    }

    // TODO: to make it null safety
    operator fun get(type: Type): String? {
        return mPaths[type]
    }

    fun setCleanCallback(cb: (() -> Unit)?) {
        mCleanCallback = cb
    }

    /**
     * Call this to clean extracted files.
     */
    fun clean() = mCleanCallback?.invoke()

    override fun toString(): String {
        val sb = StringBuilder()
        for (t in Type.values()) {
            sb.append(String.format("Type: %d, path: ", t, mPaths[t]))
        }
        return sb.toString()
    }
}
